lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -2,47 +2,217 @@
+    # point to actually extract.
+    #
+    # The planning phase is similar to the extraction phase. Though in the
+    # planning phase, we can ignore all the actual attribute values and only
+    # look at the attribute names. This way, we can compute the attribute names
+    # for each set of dependencies without populating their actual values.
+    #
+    # For example, in the planning phase,
+    #
+    #   {a:nil} x {b: nil} => {a: nil, b: nil}
+    #   {a:nil, b:nil} x {a: nil: b: nil} => {a: nil, b: nil}
+    #
+    # and in the extraction phase, that's where the # of dependency can
+    # actually grow significantly:
+    #
+    #   {a: [1,2,3]} x {b: [1,2,3]} => [{a: 1, b: 1}, ....]
+    #   {a:[1,2], b:[1,2]} x {a: [1,2,3]: b: [1,2,3]} => [{a: 1, b: 1}, ...]
+    #
+    class Plan
+      class DependencySizeEstimation
+        def initialize(hash = nil)
+          @hash = hash
+        end
+
+        def +(other)
+          merged_hash = hash.dup
+          other.hash.each do |k, v|
+            merged_hash[k] += v
+          end
+          self.class.new(merged_hash)
+        end
+
+        def *(other)
+          merged_hash = hash.dup
+          other.hash.each do |k, v|
+            if merged_hash.include?(k)
+              merged_hash[k] *= v
+            else
+              merged_hash[k] = v
+            end
+          end
+          self.class.new(merged_hash)
+        end
+
+        def [](key)
+          hash[key]
+        end
+
+        def []=(key, val)
+          hash[key] = val
+        end
+
+        def to_i
+          ret = 0
+          hash.each do |_, v|
+            ret += v
+          end
+          ret
+        end
 
-      # The tree is almost always right-heavy. Merge into the right node for better
-      # performance.
-      other.params.merge!(params) do |_, other_attrs_set, attrs_set|
-        if other_attrs_set.empty?
-          attrs_set
-        elsif attrs_set.empty?
-          other_attrs_set
-        else
-          attrs_set + other_attrs_set
+        protected
+
+        def hash
+          @hash ||= Hash.new(0)
         end
       end
 
-      other
+      attr_accessor :dependency_size_estimation
+      attr_accessor :model_attrs
+
+      def initialize(bind_params)
+        @dependency_size_estimation = DependencySizeEstimation.new
+        @model_attrs = Hash.new do |models, model|
+          models[model] = Set.new
+        end
+
+        # An aggregated bind_params node can only obtain params by combining
+        # its children nodes
+        return if !bind_params.__send__(:operator).nil?
+
+        bind_params.params.each do |model, attrs_set|
+          @dependency_size_estimation[model] += attrs_set.size
+          attrs_set.each do |attrs|
+            # [k, nil]: Ignore the attr value and keep the name only
+            @model_attrs[model] << attrs.keys.map { |k| [k, nil] }.to_h
+          end
+        end
+      end
```

Comment:
separate question from the code: do we have an estimate here how much time will take to compute the plan dependency size? how does it scale when the attr set is large.

Reply:
Well, the attrs_set can never be that big. that's bounded by the length of different columns in the SQL query.  Calculating plan dependency size is an O(n) operation where n is the length of the SQL query